### PR TITLE
Optimize target group operations

### PR DESCRIPTION
**Pull request by solmonk**
**What type of PR is this?**

enhancement

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

* Add caching to ListTagsForResource calls, we may expand this to other calls as well
* Optimize list operation of target group
  * fetch tags at once using GetResources() api
  * Do not make unnecessary get() calls, as tgSummary is sufficient in most cases
* Add ProtocolVersion on TG tags, this saves multiple GetTargetGroup() calls for TG cleanup.
* Optimize find operation of target group
  * Utilizes GetResources() api to filter out target group

**If an issue # is not available please add repro steps and logs from aws-gateway-controller showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this PR introduce any new dependencies?**:

**Will this break upgrades or downgrades. Has updating a running cluster been tested?**:

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/aws/services/tagging.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	maxArnsPerGetResourcesApi = 100
+)
+
+type Tags = map[string]*string
+
+type Tagging interface {
+	taggingapiiface.ResourceGroupsTaggingAPIAPI
+
+	// Receives a list of arns and returns arn-to-tags map.
+	GetTagsFromArns(ctx context.Context, arns []*string) (map[string]Tags, error)
```
Comment (mikhail-aws) [inline pkg/aws/services/tagging.go:28]: From -> For?

Comment (mikhail-aws) [reply pkg/aws/services/tagging.go:28]: arns []string, I dont think we need reference type here

Comment (mikhail-aws) [reply pkg/aws/services/tagging.go:28]: Also you can consider a strong typed version arn.ARN, much easier to use once constructed.

https://pkg.go.dev/github.com/aws/aws-sdk-go@v1.47.9/aws/arn#Parse

```go
tgArns := utils.SliceMap(resp, func(tg *vpclattice.TargetGroupSummary) arn.ARN { 
  tgArn , _ := arn.Parse(*tg.Arn)
  return tgArn
})
```

File: pkg/aws/services/tagging.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+	// Finds one resource that matches the given set of tags.
+	FindResourceWithTags(ctx context.Context, resourceType string, tags Tags) (*string, error)
```
Comment (mikhail-aws) [inline pkg/aws/services/tagging.go:31]: With -> By, also should return arn.ARN, not clear what *string means here

Comment (mikhail-aws) [reply pkg/aws/services/tagging.go:31]: resourceType string, can be strong typed, we have limited options better to keep it enum-like

File: pkg/aws/services/tagging.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	}
+	resp, err := t.GetResourcesWithContext(ctx, input)
+	if err != nil {
+		return nil, err
+	}
+	if len(resp.ResourceTagMappingList) == 0 {
+		return nil, NewNotFoundError("tag", "matching criteria")
+	}
+	// assume one result
+	return resp.ResourceTagMappingList[0].ResourceARN, nil
```
Comment (mikhail-aws) [inline pkg/aws/services/tagging.go:72]: I think assumption about how many results to expect should be on caller of this function, not here. So returning a slice of results would be better. Also empty slice as return an easy check for not found.

File: pkg/aws/services/vpclattice.go [hunk 4]
```diff
@@ -118,9 +120,11 @@ func NewDefaultLattice(sess *session.Session, region string) *defaultLattice {
 		endpoint = latticeEndpoint
 	}
 
-	latticeSess = vpclattice.New(sess, aws.NewConfig().WithRegion(region).WithEndpoint(endpoint).WithMaxRetries(20))
+	latticeSess := vpclattice.New(sess, aws.NewConfig().WithRegion(region).WithEndpoint(endpoint).WithMaxRetries(20))
```
Comment (mikhail-aws) [inline pkg/aws/services/vpclattice.go:123]: what do you think about 20 retries and it's impact on throttling/performance?

File: pkg/aws/services/vpclattice.go [hunk 5]
```diff
@@ -224,6 +228,36 @@ func (d *defaultLattice) ListTargetGroupsAsList(ctx context.Context, input *vpcl
...
+	}
+	return out, nil
+}
+
+func tagCacheKey(arn string) string {
+	return "tag-" + arn
+}
+
+func (d *defaultLattice) TagResourceWithContext(ctx context.Context, input *vpclattice.TagResourceInput, option ...request.Option) (*vpclattice.TagResourceOutput, error) {
+	if d.cache != nil {
```
Comment (mikhail-aws) [inline pkg/aws/services/vpclattice.go:254]: why cache should be nil?

File: pkg/deploy/lattice/target_group_manager.go [hunk 6]
```diff
@@ -238,98 +237,98 @@ func (s *defaultTargetGroupManager) Delete(ctx context.Context, modelTg *model.T
...
-		if status == vpclattice.TargetGroupStatusCreateFailed {
-			continue
+	latticeTg, err := s.cloud.Lattice().GetTargetGroupWithContext(ctx, &vpclattice.GetTargetGroupInput{
+		TargetGroupIdentifier: arn,
+	})
+	if err != nil {
+		if services.IsLatticeAPINotFoundErr(err) {
+			return nil, nil
 		}
+		return nil, err
```
Comment (mikhail-aws) [inline pkg/deploy/lattice/target_group_manager.go:294]: nit:

```go
return nil, services.IgnoreNotFound(err)
```

File: pkg/deploy/lattice/target_group_manager.go [hunk 7]
```diff
@@ -238,98 +237,98 @@ func (s *defaultTargetGroupManager) Delete(ctx context.Context, modelTg *model.T
...
-			case vpclattice.TargetGroupStatusDeleteFailed:
-				return latticeTg, nil
-			case vpclattice.TargetGroupStatusDeleteInProgress:
-				return nil, errors.New(LATTICE_RETRY)
-			}
+	// we ignore create failed status, so may as well check for it first
+	status := aws.StringValue(latticeTg.Status)
+	if status == vpclattice.TargetGroupStatusCreateFailed {
+		return nil, nil
+	}
```
Comment (mikhail-aws) [inline pkg/deploy/lattice/target_group_manager.go:301]: I dont follow why "findTargetGroup" make decisions about status and retry errors. It's a surprising behavior with "find" function. 

May be name is misleading? But it seems too much of responsibility for this function: searching, discarding "some" results, matching, retrying.

**Review by mikhail-aws — CHANGES_REQUESTED** [review]

File: pkg/aws/services/tagging.go (hunk 1 above)
Comment (solmonk) [reply pkg/aws/services/tagging.go:28]: I would prefer string more, ARN is purely an identifier here and processing it adds no value.

File: pkg/aws/services/tagging.go (hunk 2 above)
Comment (solmonk) [reply pkg/aws/services/tagging.go:31]: makes sense

File: pkg/aws/services/tagging.go (hunk 3 above)
Comment (solmonk) [reply pkg/aws/services/tagging.go:72]: makes sense

File: pkg/aws/services/vpclattice.go (hunk 4 above)
Comment (solmonk) [reply pkg/aws/services/vpclattice.go:123]: I have never observed backoffs getting more than 2~3 on manual tests, so it is hard for me to think that it is affecting anything. we can definitely revisit this later though.

File: pkg/aws/services/vpclattice.go (hunk 5 above)
Comment (solmonk) [reply pkg/aws/services/vpclattice.go:254]: I think when using cache we should almost always have a way to switch off for testing purpose, etc.

File: pkg/deploy/lattice/target_group_manager.go (hunk 6 above)
Comment (solmonk) [reply pkg/deploy/lattice/target_group_manager.go:294]: makes sense

File: pkg/deploy/lattice/target_group_manager.go (hunk 7 above)
Comment (solmonk) [reply pkg/deploy/lattice/target_group_manager.go:301]: Discarding some status should make total sense. It is always easier to think that `CREATE_FAILED` target group doesn't exist, and `DELETE_FAILED` target group just exists.

On the other hand, `LATTICE_RETRY` is currently like a magical fallback for us to "hold off" decision whenever we would like. It makes sense if you argue we are abusing it too much, but I find it aligning with other controller behaviors so didn't decide to touch yet.

File: pkg/aws/services/tagging.go [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+//go:generate mockgen -destination tagging_mocks.go -package services github.com/aws/aws-application-networking-k8s/pkg/aws/services Tagging
+
+type ResourceType string
+
+const (
+	resourceTypePrefix = "vpc-lattice:"
+
+	ResourceTypeTargetGroup ResourceType = resourceTypePrefix + "targetgroup"
+
+	maxArnsPerGetResourcesApi = 100
```
Comment (mikhail-aws) [inline pkg/aws/services/tagging.go:21]: nit: can you add link to the limit please

File: pkg/aws/services/vpclattice.go (hunk 5 above)
Comment (mikhail-aws) [reply pkg/aws/services/vpclattice.go:254]: in this case should be interface with no-op cache

File: pkg/deploy/lattice/target_group_manager.go [hunk 9]
```diff
@@ -124,19 +123,19 @@ func (s *defaultTargetGroupManager) update(ctx context.Context, targetGroup *mod
...
 	_, err := s.cloud.Lattice().UpdateTargetGroupWithContext(ctx, &vpclattice.UpdateTargetGroupInput{
 		HealthCheck:           healthCheckConfig,
-		TargetGroupIdentifier: latticeTgSummary.Id,
+		TargetGroupIdentifier: latticeTg.Id,
 	})
 
 	if err != nil {
 		return model.TargetGroupStatus{},
-			fmt.Errorf("Failed UpdateTargetGroup %s due to %s", aws.StringValue(latticeTgSummary.Id), err)
+			fmt.Errorf("Failed UpdateTargetGroup %s due to %s", aws.StringValue(latticeTg.Id), err)
```
Comment (mikhail-aws) [inline pkg/deploy/lattice/target_group_manager.go:131]: nit: %w for err

File: pkg/deploy/lattice/target_group_manager.go [hunk 10]
```diff
@@ -238,98 +237,99 @@ func (s *defaultTargetGroupManager) Delete(ctx context.Context, modelTg *model.T
...
+	arns, err := s.cloud.Tagging().FindResourcesByTags(ctx, services.ResourceTypeTargetGroup,
+		model.TagsFromTGTagFields(modelTargetGroup.Spec.TargetGroupTagFields))
 	if err != nil {
 		return nil, err
 	}
+	if len(arns) == 0 {
+		return nil, nil
+	}
+	// Can get only one arn through the above search criteria.
+	arn := arns[0]
```
Comment (mikhail-aws) [inline pkg/deploy/lattice/target_group_manager.go:289]: lets throw or log error when assumption does not work

File: pkg/deploy/lattice/target_group_manager.go [hunk 11]
```diff
@@ -238,98 +237,99 @@ func (s *defaultTargetGroupManager) Delete(ctx context.Context, modelTg *model.T
...
+		Type:          latticeTg.Type,
+		VpcIdentifier: latticeTg.Config.VpcIdentifier,
+	}, nil) // we already know that tags match
+	if err != nil {
+		return nil, err
+	}
+	if match {
+		switch status {
+		case vpclattice.TargetGroupStatusCreateInProgress, vpclattice.TargetGroupStatusDeleteInProgress:
+			return nil, errors.New(LATTICE_RETRY)
```
Comment (mikhail-aws) [inline pkg/deploy/lattice/target_group_manager.go:319]: ```retrun fmt.Errorf("%w, delete in progress id=%d", runtime.RetryErr)```

**Review by mikhail-aws — APPROVED** [review]

**Pull request merged**